lib/iris/coords.py

```diff
@@ -143,6 +143,147 @@ def __new__(cls, name_or_coord, minimum, maximum,
+    #         i    i+1
+    #    j    @0    @1
+    #  j+1    @3    @2
+    def mod360_diff(x1, x2):
+        diff = x1 - x2
+        diff = (diff + 360.0 + 180.0) % 360.0 - 180.0
+        return diff
+
+    # Check contiguity for 2 bounds per point or 4 bounds per point only.
+    if bds.shape[2] not in [2, 4]:
+        raise ValueError('2D coordinates must have 2 or 4 bounds per point '
+                         'for 2D coordinate plotting')
+    elif bds.shape[2] == 2:
+        # NOTE: the bounds supplied are only the ones relevent to the
+        # coordinate, i.e. bounds on longitude coord will need to be checked
+        # in the x-direction (see x_diffs), whereas bounds on latitude coord
+        # will need to be checked in y-direction (see y_diffs)
+        if axis == 'X':
+            # Compare cell with the cell next to it (i+1)
+            diffs = mod360_diff(bds[:, :-1, 1], bds[:, 1:, 0])
+        elif axis == 'Y':
+            # Compare cell with the cell above it (j+1)
+            diffs = mod360_diff(bds[:-1, :, 1], bds[1:, :, 0])
+        else:
+            raise ValueError('Coordinate does not map to X or Y dimension; '
+                             'unable to check for contiguity.')
+    elif bds.shape[2] == 4:
+        if axis == 'X':
+            diffs = mod360_diff(bds[:, :-1, 1], bds[:, 1:, 0])
+        elif axis == 'Y':
+            diffs = mod360_diff(bds[:-1, :, 3], bds[1:, :, 0])
+        else:
+            raise ValueError('Coordinate does not map to X or Y dimension; '
+                             'unable to check for contiguity.')
+
+    # At this point, diffs is an array of differences between bounds in one
+    # direction only.  It is unnecessary and nonsensical to have both
+    # x-diffs and y-diffs for one coord (i.e. longitude)
+    contiguous = np.all(diffs < abs_tol)
+
+    if not contiguous:
+        # True where data exists
+        mask_invert = np.logical_not(data.mask)
+
+        # Where a discontinuity occurs the grid will not be created correctly.
+        # This does not matter if the data is masked as this is not plotted.
+        # So for places where data exists (opposite of the mask) AND a
+        # diff exists. If any exist, raise a warning
+        not_masked_at_discontinuity = np.any(np.logical_and
+                                             (mask_invert[:, :-1], diffs))
+
+        # If discontinuity occurs but not masked, any grid will be created
+        #  incorrectly, so raise a warning
+        if not_masked_at_discontinuity:
+            raise ValueError('The bounds of the {} coordinate are not'
+                             ' contiguous and data is not masked where the'
+                             ' discontiguity occurs. Not able to create a'
+                             ' suitable mesh to give to'
+                             ' Matplotlib'.format(coord.name()))
+        else:
+            warnings.warn('The bounds of the {} coordinate are not contiguous.'
```

Comment:
I'm not keen on issuing a warning in these cases, as we have seen that some of the well-known testcase data will  *always* give this warning, and there is no practical change you can make to stop that. 
IMHO a warning is simply not useful unless there is an obvious code change that you can make (and should make) to acknowledge and/or fix a problem  -- ***ideally*** the warning text will suggest appropriate action. 
Otherwise it is just more noise !

You could potentially add a keyword here, which the user can use to acknowledge the problem + silence the warning.  Not really sure whether I think that is a good idea or not  in this case ...